### ISSUE TITLE
Incorrect error printed by `rustc` when using generics in constants

### ISSUE THREAD

**Issue by InnocentusLime** [rust-lang/rust#82623]
I tried the following piece of code:

```rust
use std::marker::PhantomData;

trait Padded {
    const PAD : usize;
}

struct Test<T : Padded> 
{
    _marker : PhantomData<T>,
    _pad : [u8; <T as Padded>::PAD],
}

fn main() {}
```
What I expected `rustc 1.50.0` to do is abort with an error telling that generic parameters are not yet allowed in constant expressions. The stable version of the compiler, however, prints out the following error, making it seem like it's ignoring the trait bound:

```
error[E0277]: the trait bound `T: Padded` is not satisfied
  --> src/main.rs:10:17
   |
4  |     const PAD : usize;
   |     ------------------ required by `Padded::PAD`
...
10 |     _pad : [u8; <T as Padded>::PAD],
   |                 ^^^^^^^^^^^^^^^^^^ the trait `Padded` is not implemented for `T`
   |
help: consider further restricting this bound
   |
7  | struct Test<T : Padded + Padded> 
   |                        ^^^^^^^^
```

The `beta` and `nightly` versions of the compiler report the error correctly:

```
>error: generic parameters may not be used in const operations
 --> src/main.rs:10:18
   |
10 |     _pad : [u8; <T as Padded>::PAD],
   |                  ^ cannot perform const operation using `T`
   |
   = note: type parameters may not be used in const expressions
   = help: use `#![feature(const_generics)]` and `#![feature(const_evaluatable_checked)]` to allow generic const expressions
```

**Comment by jonas-schievink** [comment]
Fixed in beta, so closing
